docs/development/pigs/pig-009.rst

```diff
@@ -0,0 +1,231 @@
+
+For this reason we favour the inverse CDF sampling method, as a simple to implement
+and general sampling method.
+The precision of the inverse CDF sampling method can be controlled by the resolution
+of the input probability mass function (PMF) and is in practice only limited by the
+available memory. We will study the required bin-size of the PMFs to reach sufficient
+precision. If we find the inverse CDF sampling method to be not precise enough, it is 
+still possible to achieve better precision adopting the rejection sampling. This will 
+not have a strong impact on the structure of the event-sampler.
+
+
+Proposal
+========
+
+We propose to include in `gammapy.cube` an high level interface (HLI) class, labelled as
+`MapDatasetEventSampler`. This class handles the complete event sampling process, 
+including the corrections related to the IRF and source temporal variability, for a given 
+GTI / observation.
+
+As a first step, the User should create a `MapDataset` object, as described below:
+
+# 20 lines of code with the `MapDataset` object setup.
+
+This object will contain the pointing sky coordinates, the GTI, and the setup of all the 
+models (spectra, spatial morphology, temporal model) for any given source, and it is passed 
+as input parameter to the `MapDatasetEventSampler`. It is important to note that the 
+`MapDataset` object can store information for more than one source. Then, a `.sample` method 
+will draw the sampled events and will provide an output `~astropy.table.Table` object. The 
+latter will containt the reconstructed sky positions, energies, times, and an `EVENT_ID` and 
+`MC_ID`. `EVENT_ID` is a unique number or a string to identify the sampled event, while 
+`MC_ID` is a unique ID (number or string) to identify the model component the event was sampled
+from. The `MapDatasetSampler` should also fill the mandatory header information for event list 
+files decribed on `gamma-astro-data-formats`_. The `MapDatasetSampler` will be designed to work 
+alternatively for input flux or counts maps.
+
+With the code example above, this could go as:
+
+::
+    
+    dataset = MapDataset([pointing, GTI, models])
+    dataset_sampler = MapDatasetEventSampler(dataset, random_state)
+    event_list = dataset_sampler.sample()
+    
+where `random_state` is the seed for the random number generator. 
+
+
+MapDatasetEventSampler
+----------------------
+
+The general design of the `sample` method is as follows:
+
+::
+
+    def sample(dataset, random_state)
+       """Sample events from a `MapDataset`"""
+
+       events_list = []
+    
+       for evaluator in dataset.evaluators:
+            npred = evaluator.compute_npred()
+            n_events = random_state.poisson(npred.data.sum())
+            events = npred.sample(n_events, random_state)
+            events_list.append(events)
+    
+       events_src = vstack(events_list)
+       events_src = dataset.psf.distribute(events_src, random_state)
```

Comment:
Suggest to name this method also "sample", not "distribute".
It samples the EDISP or PSF distribution functions, it is a kind of sampling.
I just think if all sampling-related code has the same name it's easier to grok and find.

Reply:
+1

Reply:
Done